internal/app/go-filecoin/storage_miner_connector/connector.go

```diff
@@ -303,64 +305,67 @@ func (m *StorageMinerNodeConnector) GetSealTicket(ctx context.Context) (storage.
 // responsible for choosing an interval-value, which is a quantity of blocks to
 // wait (after the block in which the pre-commit message is mined) before
 // computing and sampling a seed.
-func (m *StorageMinerNodeConnector) GetSealSeed(ctx context.Context, preCommitMsg cid.Cid, interval uint64) (seed <-chan storage.SealSeed, err <-chan error, invalidated <-chan struct{}, done <-chan struct{}) {
-	sc := make(chan storage.SealSeed)
-	hc := make(chan block.TipSetKey)
-	ec := make(chan error)
-	ic := make(chan struct{})
-	dc := make(chan struct{})
+func (m *StorageMinerNodeConnector) GetSealSeed(ctx context.Context, preCommitMsg cid.Cid, interval uint64) (<-chan storage.SealSeed, <-chan error, <-chan struct{}, <-chan struct{}) {
+	seedCh := make(chan storage.SealSeed)
+	errCh := make(chan error)
+	invalidCh := make(chan struct{})
+	doneCh := make(chan struct{})
 
 	go func() {
 		h, exitCode, err := m.waitForMessageHeight(ctx, preCommitMsg)
 		if err != nil {
-			ec <- err
+			errCh <- err
 			return
 		}
 
 		if exitCode != 0 {
-			ec <- xerrors.Errorf("non-zero exit code for pre-commit message %d", exitCode)
+			errCh <- xerrors.Errorf("non-zero exit code for pre-commit message %d", exitCode)
 			return
 		}
 
-		m.chainHeightScheduler.AddListener(h + interval)
-	}()
+		listener := m.chainHeightScheduler.AddListener(h + interval)
 
-	// translate tipset key to seal seed handler
-	go func() {
+		// translate tipset key to seal seed handler
 		for {
 			select {
-			case key := <-hc:
+			case key := <-listener.HitCh:
 				ts, err := m.chainState.GetTipSet(key)
 				if err != nil {
-					ec <- err
-					break
+					listener.ErrCh <- err
+					continue
 				}
 
 				tsHeight, err := ts.Height()
 				if err != nil {
-					ec <- err
-					break
+					listener.ErrCh <- err
+					continue
 				}
 
 				randomness, err := m.chainState.SampleRandomness(ctx, types.NewBlockHeight(tsHeight))
 				if err != nil {
-					ec <- err
-					break
+					listener.ErrCh <- err
+					continue
 				}
 
-				sc <- storage.SealSeed{
+				seedCh <- storage.SealSeed{
 					BlockHeight: tsHeight,
 					TicketBytes: randomness,
 				}
-			case <-dc:
+			case err := <-listener.ErrCh:
```

Comment:
Is some other goroutine listening on this and is that a problem?

Reply:
I think what you might want to do is return the listteners err, invalid and done channels, remove these cases (keeping the ctx.Done() one).

Reply:
The idea is that we're mapping all the channels from the listener to channels created in this method. This this method use the errCh as well as the listener. We can't return the listeners invalidCh and doneCh because this method should return immediately and those channels aren't available until after the wait.